File: src/visual-graph/prune/bfs.ts
export function bfs(
  starts: ReadonlySet<string>,
  adj: ReadonlyMap<string, readonly string[]>,
  maxDepth: number,
): Set<string> {
  const reached = new Set<string>(starts);
  if (maxDepth <= 0) {
    return reached;
  }
  let frontier = new Set<string>(starts);
  for (let depth = 0; depth < maxDepth && frontier.size > 0; depth++) {
    const next = new Set<string>();
    for (const id of frontier) {
      const neighbors = adj.get(id);
      if (neighbors === undefined) {
        continue;
      }
      for (const n of neighbors) {
        if (!reached.has(n)) {
          reached.add(n);
          next.add(n);
        }
      }
    }
    frontier = next;
  }
  return reached;
}

File: crates/unsnarl-visual-graph/src/prune/bfs.rs
//! Breadth-first traversal helper for pruning.
//!
//! Mirrors `ts/src/visual-graph/prune/bfs.ts`.

use std::collections::{HashMap, HashSet};

pub fn bfs(
    starts: &HashSet<String>,
    adj: &HashMap<String, Vec<String>>,
    max_depth: i32,
) -> HashSet<String> {
    let mut reached: HashSet<String> = starts.clone();
    if max_depth <= 0 {
        return reached;
    }
    let mut frontier: HashSet<String> = starts.clone();
    let mut depth = 0;
    while depth < max_depth && !frontier.is_empty() {
        let mut next: HashSet<String> = HashSet::new();
        for id in &frontier {
            if let Some(neighbors) = adj.get(id) {
                for n in neighbors {
                    if !reached.contains(n) {
                        reached.insert(n.clone());
                        next.insert(n.clone());
                    }
                }
            }
        }
        frontier = next;
        depth += 1;
    }
    reached
}

#[cfg(test)]
#[path = "bfs_test.rs"]
mod bfs_test;
